docs/mempools/dd.md

```diff
@@ -0,0 +1,106 @@
+- Geth checks that txs have a minimum gas price before being accepted. For replacements (ie two txs with same sender and nonce), it checks that price bumps are at least of a given %.
+  - `PriceLimit` Minimum gas price to enforce for acceptance into the pool: 1
+  - `PriceBump` Minimum price bump percentage to replace an already existing transaction by nonce: 10%
+- When adding a new tx to the pool, after static validations, geth enqueues the tx as non-executable, and waits for the loop to promote it.
+  - [If the tx pool is full](https://github.com/ethereum/go-ethereum/blob/80b8d7a13c20254a9cfb9f7cbca1ab00aa6a3b50/core/txpool/legacypool/legacypool.go#L691-L692), it discards cheaper txs based on gas tip (ie priority fees). Only the global slots seem to be considered here.
+- When cleaning up the pool in a loop, loops over pending (executable?) txs for every sender that has gone over AccountSlots, and drops txs (based on nonce) form them. Also loops over future (non-executable?) txs based on **heartbeats**: accounts with the most time without any activity get their txs pruned first.
+
+## Difficulties
+
+In addition to all complications that Ethereum has, we also have the issue that a tx public execution can invalidate an arbitrary number of existing txs just by emitting nullifiers. We have no way of knowing that in advance.
+
+Also, while for Ethereum a "replacement" is just a tx with the same nonce and sender as an existing one, for us any tx that shares a nullifier can technically be a replacement. This also means that tx A may be a replacement for B and C, but B and C may be unrelated to each other.
+
+Also, while our `fee_payer` slightly matches Ethereum's `sender`, it's possible that many users (if not all) use the same very few fee payers (in our case, FPCs), so there is likely no point in setting limits per sender as Ethereum does. Remember that, thanks to privacy, we cannot know the sender of a tx. On the flip side, we know that two txs do come from the same user if they share a private-land nullifier.
+
+## Design
+
+To recap, we need to consider:
+
+- Balance of fee-payers
+- Conflicting nullifiers
+- Max block number
+- Gas fees and limit vs current base fees and limits
+- Archive root (only on reorgs)
+
+We propose keeping the following indices for all txs:
+
+- priority fee
+- fee-payer
+- nullifier
+- base fee
+- gas limit
+- max block number
+
+When adding a tx, in addition to running all static checks:
+
+- We check if the current balance of the fee payer, minus the max cost of all pending txs for that fee payer, is enough to pay for this tx. If it is not, we try evicting other txs with a lower priority fee. If that works, and all other checks pass, we include the tx dropping the others.
+- We check if it shares a nullifier with any existing pending tx. If it pays more than all of the conflicting ones, and it passes all other checks, we include it and drop the other ones.
```

Comment:
Or nullifier already in the state ye? Earlier alluded to duplicates, just adding explicit comment here.